File: src/main/java/net/skyenetwork/skyecrates/particles/ParticleManager.java
package net.skyenetwork.skyecrates.particles;

import net.skyenetwork.skyecrates.SkyeCratesPlugin;
import net.skyenetwork.skyecrates.crates.CrateManager;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.scheduler.BukkitTask;

import java.util.Map;

public class ParticleManager {
    
    private final SkyeCratesPlugin plugin;
    private BukkitTask particleTask;
    
    public ParticleManager(SkyeCratesPlugin plugin) {
        this.plugin = plugin;
    }
    
    public void start() {
        stop(); // Stop any existing task
        
        particleTask = Bukkit.getScheduler().runTaskTimer(plugin, this::spawnParticles, 
                0L, plugin.getConfigManager().getParticleInterval());
    }
    
    public void stop() {
        if (particleTask != null) {
            particleTask.cancel();
            particleTask = null;
        }
    }
    
    public void reload() {
        start(); // Restart with new interval
    }
    
    private void spawnParticles() {
        Map<Location, String> placedCrates = plugin.getCrateManager().getPlacedCrates();
        
        for (Map.Entry<Location, String> entry : placedCrates.entrySet()) {
            Location location = entry.getKey();
            String crateName = entry.getValue();
            
            CrateManager.CrateConfig crate = plugin.getCrateManager().getCrate(crateName);
            if (crate == null) {
                continue;
            }
            
            CrateManager.ParticleConfig particles = crate.getParticles();
            
            // Spawn particles above the crate
            Location particleLocation = location.clone().add(0.5, 1.5, 0.5);
            
            if (plugin.getConfigManager().isForceParticles()) {
                // Force particles to all players regardless of their settings
                if (location.getWorld() != null) {
                    location.getWorld().spawnParticle(
                        particles.getType(),
                        particleLocation,
                        particles.getCount(),
                        particles.getDeltaX(),
                        particles.getDeltaY(),
                        particles.getDeltaZ(),
                        particles.getSpeed(),
                        null, // data
                        true  // force - ignores player particle settings
                    );
                }
            } else {
                // Normal particles that respect player settings
                if (location.getWorld() != null) {
                    location.getWorld().spawnParticle(
                        particles.getType(),
                        particleLocation,
                        particles.getCount(),
                        particles.getDeltaX(),
                        particles.getDeltaY(),
                        particles.getDeltaZ(),
                        particles.getSpeed()
                    );
                }
            }
        }
    }
}

File: src/main/kotlin/net/skyenetwork/skyecrates/particles/ParticleManager.kt
package net.skyenetwork.skyecrates.particles

import net.skyenetwork.skyecrates.SkyeCratesPlugin
import org.bukkit.Bukkit
import org.bukkit.scheduler.BukkitTask

class ParticleManager(private val plugin: SkyeCratesPlugin) {
    
    private var particleTask: BukkitTask? = null
    
    fun start() {
        stop() // Stop any existing task
        
        particleTask = Bukkit.getScheduler().runTaskTimer(plugin, {
            spawnParticles()
        }, 0L, plugin.configManager.particleInterval)
    }
    
    fun stop() {
        particleTask?.cancel()
        particleTask = null
    }
    
    fun reload() {
        start() // Restart with new interval
    }
    
    private fun spawnParticles() {
        plugin.crateManager.getPlacedCrates().forEach { (location, crateName) ->
            val crate = plugin.crateManager.getCrate(crateName) ?: return@forEach
            val particles = crate.particles
            
            // Spawn particles above the crate
            val particleLocation = location.clone().add(0.5, 1.5, 0.5)
            
            if (plugin.configManager.forceParticles) {
                // Force particles to all players regardless of their settings
                location.world?.spawnParticle(
                    particles.type,
                    particleLocation,
                    particles.count,
                    particles.deltaX,
                    particles.deltaY,
                    particles.deltaZ,
                    particles.speed,
                    null, // data
                    true  // force - ignores player particle settings
                )
            } else {
                // Normal particles that respect player settings
                location.world?.spawnParticle(
                    particles.type,
                    particleLocation,
                    particles.count,
                    particles.deltaX,
                    particles.deltaY,
                    particles.deltaZ,
                    particles.speed
                )
            }
        }
    }
}
